File: typescript/src/20.valid-parentheses.ts
/*
 * @lc app=leetcode id=20 lang=typescript
 *
 * [20] Valid Parentheses
 */

// @lc code=start
function isValid(s: string): boolean {
  let stack: string[] = [];
  for (const element of s) {
    if (element === "(") stack.push(")");
    else if (element === "[") stack.push("]");
    else if (element === "{") stack.push("}");
    else if (!stack.length || stack.pop() !== element) return false;
  }

  return !stack.length;
}
// @lc code=end

File: rust/src/bin/20.valid-parentheses.rs
/*
 * @lc app=leetcode id=20 lang=rust
 *
 * [20] Valid Parentheses
 */

use std::vec;

// @lc code=start
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack: Vec<char> = Vec::new();
        for element in s.chars() {
            if element == '(' {
                stack.push(')');
            }
            else if element=='[' {
                stack.push(']');
            }
            else if element=='{' {
                stack.push('}');
            }
            else if stack.is_empty() || stack.pop().unwrap() != element {
                return false;
            }
        }
        stack.is_empty()
    }
}
// @lc code=end

